src/common/socket.cpp

```diff
@@ -1173,75 +1171,48 @@ static int32 connect_check_(uint32 ip)
 		break;
 	}
 
-	// Inspect connection history
-	while( hist ) {
-		if( ip == hist->ip )
-		{// IP found
-			if( hist->ddos )
-			{// flagged as DDoS
-				return (connect_ok == 2 ? 1 : 0);
-			} else if( DIFF_TICK(gettick(),hist->tick) < ddos_interval )
-			{// connection within ddos_interval
-				hist->tick = gettick();
-				if( hist->count++ >= ddos_count )
-				{// DDoS attack detected
-					hist->ddos = 1;
-					ShowWarning("connect_check: DDoS Attack detected from %d.%d.%d.%d!\n", CONVIP(ip));
-					return (connect_ok == 2 ? 1 : 0);
-				}
-				return connect_ok;
-			} else
-			{// not within ddos_interval, clear data
-				hist->tick  = gettick();
-				hist->count = 0;
-				return connect_ok;
-			}
+	auto it = connectHistoryMap.find(ip); 
+	if (it == connectHistoryMap.end()) {
+		connectHistoryMap.emplace(ip, gettick());
+		return connect_ok;
+	}
+	if (it->second.ddos_) {
+		return (connect_ok == 2 ? 1 : 0);
+	}
+	if (DIFF_TICK(gettick(), it->second.tick_) < ddos_interval) {
+		it->second.tick_ = gettick();
+		if (it->second.count_++ >= ddos_count) {
+			it->second.ddos_ = true;
+			ShowWarning("connect_check: DDoS Attack detected from %d.%d.%d.%d!\n", CONVIP(ip));
+			return (connect_ok == 2 ? 1 : 0);
 		}
-		hist = hist->next;
-	}
-	// IP not found, add to history
-	CREATE(hist, ConnectHistory, 1);
-	memset(hist, 0, sizeof(ConnectHistory));
-	hist->ip   = ip;
-	hist->tick = gettick();
-	hist->next = connect_history[ip&0xFFFF];
-	connect_history[ip&0xFFFF] = hist;
+	} else {
+		it->second.tick_ = gettick();
+		it->second.count_ = 0;
+	}
 	return connect_ok;
 }
 
 /// Timer function.
 /// Deletes old connection history records.
 static TIMER_FUNC(connect_check_clear){
-	int32 i;
+	int32 previousSize = connectHistoryMap.size();
 	int32 clear = 0;
-	int32 list  = 0;
-	ConnectHistory root;
-	ConnectHistory* prev_hist;
-	ConnectHistory* hist;
-
-	for( i=0; i < 0x10000 ; ++i ){
-		prev_hist = &root;
-		root.next = hist = connect_history[i];
-		while( hist ){
-			if( (!hist->ddos && DIFF_TICK(tick,hist->tick) > ddos_interval*3) ||
-					(hist->ddos && DIFF_TICK(tick,hist->tick) > ddos_autoreset) )
-			{// Remove connection history
-				prev_hist->next = hist->next;
-				aFree(hist);
-				hist = prev_hist->next;
-				clear++;
-			} else {
-				prev_hist = hist;
-				hist = hist->next;
-			}
-			list++;
+
+	for (auto it = connectHistoryMap.begin(); it != connectHistoryMap.end(); ) {
```

Comment:
We could convert this to a ranged for loop and not really erase the entry and just set the ddos flag to false.